### PR TITLE
V2.17.0

### PR DESCRIPTION
**Pull request by cbouss**
tensorflow 2.17.0

**Destination channel:** main

### Links

- https://github.com/tensorflow/tensorflow/tree/v2.17.0
- Relevant dependency PRs:
  - https://github.com/AnacondaRecipes/python-flatbuffers-feedstock/pull/1
  - https://github.com/AnacondaRecipes/flatbuffers-feedstock/pull/4
  - https://github.com/AnacondaRecipes/tensorboard-feedstock/pull/1
  - https://github.com/AnacondaRecipes/ml_dtypes-feedstock/pull/3
  - https://github.com/AnacondaRecipes/libml_dtypes-headers-feedstock/pull/1
  - https://github.com/AnacondaRecipes/rsync-feedstock/pull/2 (for complex file copies in build.sh) 
  - https://github.com/AnacondaRecipes/snappy-feedstock/pull/5 (with rtti enabled, for linux builds)
  - https://github.com/AnacondaRecipes/keras-feedstock/pull/4 (keras >=3.2)
  - https://github.com/AnacondaRecipes/bazel-feedstock/pull/19

### Explanation of changes:
- Moving back recipe from tensorflow_recipes to here.
- Sync with CF.
- abs.yaml configured to not build cuda on CI
- Building with XNN option xnn_enable_avx512fp16=false as GCC 11.8 does not support avx512fp16
- Building with jpeg, not libturbo_jpeg
- Updated keras required.
- Snappy with rtti required on linux.
- Updated bazel to 6.5
- Updated bazel toolchain scripts based on bazel-toolchain 0.2.0 (See https://github.com/AnacondaRecipes/bazel-toolchain-feedstock/pull/2)
- Make python handling less hermetic - packages are still downloaded from pypi at build time, but our packages are used at run time. This results in a number of overdepending warnings.

### Platforms working

- [x] linux-64
- [x] linux-64 cuda
- [x] linux-aarch64
- [x] osx-64
- [x] osx-arm64
- [ ] win-64 - to be done on subsequent PR.

### CODE REVIEW

**Comment by danpetry** [comment]
The cuda packages are built, can provide them when ready to upload rest of stuff.

**Comment by JeanChristopheMorinPerso** [comment]
Would it be possible to split up the conda-forge sync from the rest of the changes? It would make the whole review a lot easier.

**Comment by cbouss** [comment]
> Would it be possible to split up the conda-forge sync from the rest of the changes? It would make the whole review a lot easier.

Sure, here is a PR with the sync changes only: https://github.com/AnacondaRecipes/tensorflow-feedstock/pull/2

File: recipe/build_pkg.sh [hunk 1]
```diff
@@ -3,7 +3,7 @@
 set -exuo pipefail
 
 # install the whl making sure to use host pip/python if cross-compiling
-${PYTHON} -m pip install --no-deps $SRC_DIR/tensorflow_pkg/*.whl
+${PYTHON} -m pip install --no-deps --no-build-isolation $SRC_DIR/tensorflow_pkg/*.whl
```
Comment (JeanChristopheMorinPerso) [inline recipe/build_pkg.sh:6]: `--no-build-isolation` is technically redundant here since wheels are not built.

Comment (cbouss) [reply recipe/build_pkg.sh:6]: Agreed but that made the linter happy and is not harmful.

File: recipe/build.sh [hunk 2]
```diff
@@ -2,6 +2,16 @@
...
+if [[ $target_platform == linux-aarch64 ]]; then
+    if [ ! -f /usr/bin/python3 ] || [ -L /usr/bin/python3 ]; then
+        rm /usr/bin/python3 || true
+        ln -s ${PYTHON} /usr/bin/python3
+    fi
+fi
+
 if [[ "$CI" == "github_actions" ]]; then
   export CPU_COUNT=4
 fi
```
Comment (danpetry) [inline recipe/build.sh:17]: libjpeg_turbo is mentioned in TF_SYSTEM_LIBS, does that need to be changed to jpeg?

File: recipe/meta.yaml [hunk 3]
```diff
@@ -103,40 +108,40 @@ requirements:
...
-    - zlib
+    - icu {{ icu }}
+    - jpeg {{ jpeg }}
+    - libpng {{ libpng }}
+    - libprotobuf {{ libprotobuf }}
+    - libprotobuf-python-headers {{ libprotobuf }}
+    - openssl {{ openssl }}
+    - pybind11 2
+    - sqlite {{ sqlite }}
+    - snappy 1.2.1
```
Comment (danpetry) [inline recipe/meta.yaml:147]:
```suggestion
    # snappy needs rtti enabled, this version of ours has it
    - snappy 1.2.1
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -191,24 +201,24 @@ outputs:
...
-        - zlib
+        - icu {{ icu }}
+        - jpeg {{ jpeg }}
+        - libcurl {{ libcurl }}
+        - libpng {{ libpng }}
+        - libprotobuf {{ libprotobuf }}
+        - libprotobuf-python-headers {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - pybind11 2
+        - snappy 1.2.1
```
Comment (danpetry) [inline recipe/meta.yaml:224]:
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -364,26 +380,26 @@ outputs:
...
+        - libabseil {{ libabseil }}
+        - giflib {{ giflib }}
+        - libgrpc {{ libgrpc }}
+        - icu {{ icu }}
+        - jpeg {{ jpeg }}
+        - libcurl {{ libcurl }}
+        - libpng {{ libpng }}
+        - libprotobuf {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - snappy 1.2.1
```
Comment (danpetry) [inline recipe/meta.yaml:407]:
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

File: recipe/meta.yaml [hunk 6]
```diff
@@ -410,28 +426,27 @@ outputs:
...
+        - giflib {{ giflib }}
+        - icu {{ icu }}
+        - libabseil {{ libabseil }}
+        - libcurl {{ libcurl }}
+        - libgrpc {{ libgrpc }}
+        - jpeg {{ jpeg }}
+        - libpng {{ libpng }}
+        - libprotobuf {{ libprotobuf }}
+        - openssl {{ openssl }}
+        - snappy 1.2.1
```
Comment (danpetry) [inline recipe/meta.yaml:455]:
```suggestion
        # snappy needs rtti enabled, this version of ours has it
        - snappy 1.2.1
```

File: recipe/custom_toolchain/BUILD [hunk 7]
```diff
@@ -7,12 +7,15 @@ filegroup(
 
 filegroup(
     name = "cc_wrapper",
-    srcs = ["cc_wrapper.sh"],
```
Comment (danpetry) [inline recipe/custom_toolchain/BUILD:10]: with this bazel toolchain stuff, it looks like this is the updated (v0.2.0) toolchain files, plus modifications that were done specifically for tensorflow. i.e. you diffed the custom_toolchain folder in conda-forge's tensorflow-feedstock with the version of bazel-feedstock it was copied from; then updated the custom_toolchain here to our latest; then re-applied these diffs? Visually, the only thing I caught extra here is the patchelf bit..?

**Review by danpetry — COMMENTED** [review]
Partial review, reviewed the items in the "explanation of changes" list, not including the one about hermetic python.

General comments:
- It would be nice to move the other tensorflow repos to the deprecated repo
- This isn't a fork of conda-forge's recipe, which we can't change now, but maybe we could add something in the README to that effect, i.e. something to tell whoever works on this next that they should pull conda-forge's changes with changes made in this PR made again on top of that.

File: recipe/custom_toolchain/BUILD (hunk 7 above)
Comment (cbouss) [reply recipe/custom_toolchain/BUILD:10]: Yes, this matches what I remember.

File: recipe/build.sh (hunk 2 above)
Comment (cbouss) [reply recipe/build.sh:17]: The tensorflow build files expect this target to be named libjpeg_turbo. The code uses the libjpeg API though (jpeglib.h), which both libjpeg and libjpeg_turbo support.

File: recipe/README.md [hunk 8]
```diff
@@ -1,3 +1,10 @@
+Recipes for Tensorflow related conda packages for tensorflow >1.8.0,<=2.12 can be found at:
+https://github.com/AnacondaRecipes/tensorflow_recipes
+
+We are moving back to using this feedstock as of tensorflow 2.17.0.
+
+
```
Comment (cbouss) [inline recipe/README.md]:
```suggestion
This isn't a fork of conda-forge's feedstock. Whoever works on this and wishes to bring changes from the conda-forge feedstock should pull conda-forge's changes and reapply changes made in this PR on top of that.
```

**Comment by cbouss** [comment]
> * It would be nice to move the other tensorflow repos to the deprecated repo

Will do once this is merged in.

> * This isn't a fork of conda-forge's recipe, which we can't change now, but maybe we could add something in the README to that effect, i.e. something to tell whoever works on this next that they should pull conda-forge's changes with changes made in this PR made again on top of that.

Added comment to README.

File: recipe/custom_toolchain/BUILD (hunk 7 above)
Comment (danpetry) [reply recipe/custom_toolchain/BUILD:10]: ok, cool. It could be nice to document this approach for whoever does the next tensorflow update. Even better, some way of using the normal toolchain files and adding in the diff somehow (maybe using sed in the build.sh script?) would be good

File: recipe/build.sh (hunk 2 above)
Comment (danpetry) [reply recipe/build.sh:17]: right. ok thanks

File: recipe/custom_toolchain/BUILD (hunk 7 above)
Comment (cbouss) [reply recipe/custom_toolchain/BUILD:10]: The way forward should be to update the normal toolchain and use it.

Comment (danpetry) [reply recipe/custom_toolchain/BUILD:10]: would that mean patchelf needs to be included in the deps for any recipes that use it? If not, is there scope for updating the normal toolchain and using it before merging this PR?

File: recipe/meta.yaml [hunk 9]
```diff
@@ -42,49 +46,50 @@ source:
...
     sha256: 2d7e100b1878084da34b5e23b49a0cbb5ee8a7add74b7dd189a82ada1cf85530
     folder: tensorflow-estimator
 
 build:
-  number: 0
-  skip: true  # [win]
+  number: {{ build }}
+  skip: true  # [win or s390x]
   skip: true  # [python_impl == 'pypy']
   skip: true  # [py<39]
```
Comment (danpetry) [inline recipe/meta.yaml]: might want to make a note that it's untested with python 3.13 in case someone tries to rebuild this for that https://www.tensorflow.org/install/source#tested_build_configurations

File: recipe/patches/0032-loosen-hermetic-python.patch [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+From: Charles Bousseau <cbousseau@anaconda.com>
+Date: Tue, 3 Sep 2024 09:01:50 -0400
+Subject: [PATCH] loosen hermetic python
+
+Tensorflow registers python dependencies using rules_python in order
+to be hermetic to system installed python packages.
+Even if packages are installed, packages are downloaded from pypi at 
+build time.
+On linux-aarch64, a number of requested packages are not available.
+This removes some declared dependencies allowing the build to proceed.
```
Comment (danpetry) [inline recipe/patches/0032-loosen-hermetic-python.patch:11]: I've been looking into this hermetic python business and haven't understood what's going on properly (although I understand the concept in general) but was wondering if you knew the following?
- if "hermetic python" is switched on, why doesn't that mean that the tf build system will download its own python and python dependencies rather than using ours? Do we want to switch hermetic python off?
- how do the bazel files provided in this feedstock interact with the bazel files in the tf build system? Are they copied over similar files in the tree or something like that?
Basically I'm curious about why removing these pypi downloads is inconsequential. I.e. how the tf build system knows to get our dependencies anyway.

**Review by danpetry — COMMENTED** [review]
I reviewed the hermetic python stuff sort of (to the best of my understanding) and still need to give a final once-over to make sure I've looked at everything.

File: recipe/meta.yaml (hunk 9 above)
Comment (cbouss) [reply recipe/meta.yaml]: As I understand, their first 3.13 version will be tensorflow 2.19. I don't think this would build with 3.13.

Comment (danpetry) [reply recipe/meta.yaml]: ok, thanks, then a skip py>312 perhaps?

File: recipe/patches/0032-loosen-hermetic-python.patch (hunk 10 above)
Comment (cbouss) [reply recipe/patches/0032-loosen-hermetic-python.patch:11]: Hermetic python is not a flag you can switch off unfortunately. Patch 0011 ensures we are using our python interpreter (which means whatever package installed in this environment is available as well). 
The tf build system does not know how about this however. Tensorflow uses https://github.com/bazelbuild/rules_python targets to configure pypi dependencies at build time (for example, `@pypi_keras//:pkg`  essentially triggers a `pip install`). A proper way would be to enhance that project to handle system dependencies instead. This is a big undertaking. 
The python dependencies are not vendored though (you can verify this by extracting the resulting tensorflow package). So at runtime, our dependencies are used as expected. 

The bazel files in the feedstock do not replace upstream files. Rather, the extra configuration is added through `.bazelrc` set in our build scripts.

Comment (danpetry) [reply recipe/patches/0032-loosen-hermetic-python.patch:11]: > at runtime, our dependencies are used as expected

and, indeed at build time?
Thanks sounds good to me.

**Review by danpetry — APPROVED** [review]
I've finished my review and am good with the changes once the builds pass

**Comment by danpetry** [comment]
We also had sporadic xcode-related failures on osx-64 and osx-arm64:
```
10/02/24 10:38:16  INFO   Analyzing: 3 targets (646 packages loaded, 25696 targets configured)                                                                  
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/local_config_apple_cc/BUILD:61:23: in                
                          apple_cc_toolchain rule @local_config_apple_cc//:cc-compiler-darwin_arm64: Xcode version must be specified to use an Apple CROSSTOOL. 
                          If your Xcode version has changed recently, verify that "xcode-select -p" is correct and then try: "bazel shutdown" to re-run Xcode   
                          configuration                                                                                                                         
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/local_config_apple_cc/BUILD:61:23: Analysis of target
                          '@local_config_apple_cc//:cc-compiler-darwin_arm64' failed                                                                            
                   INFO   ERROR: /private/var/tmp/_bazel_builder/3d27507a171c46891a410d16c736a180/external/bazel_tools/src/tools/launcher/BUILD:9:14: errors    
                          encountered resolving toolchains for @bazel_tools//src/tools/launcher:launcher                                                        
                   INFO   ERROR:                                                                                                                                
                          /private/var/folders/k1/30mswbxs7r1g6zwn8y4fyt500000gp/T/abs_fdl1h7az42/croot/tensorflow-split_1727883363694/work/tensorflow/tools/bu…
                          errors encountered resolving toolchains for //tensorflow/tools/build_info:gen_build_info
```
however restarting the build solves it. It could have been differences in configuration of our (physical) osx builders. In any case, it is using our toolchain:
```
-isystem \
    /Users/dpetry/miniconda3/envs/bld/conda-bld/tensorflow-split_1728335448941/_build_env/include/c++/v1 \
    -isystem \
    /Users/dpetry/miniconda3/envs/bld/conda-bld/tensorflow-split_1728335448941/_build_env/lib/clang/14.0.6/include \
    -isystem \
    /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/usr/include \
    -isystem \
    /Library/Developer/CommandLineTools/SDKs/MacOSX11.1.sdk/System/Library/Frameworks \
```

File: recipe/meta.yaml [hunk 11]
```diff
@@ -228,7 +243,7 @@ outputs:
         - wrapt >=1.11.0
         # TF-API needs to move in sync
         - tensorboard >=2.17,<2.18
-        - keras >=3.0
+        - keras >=3.2
       run:
         - python
         - packaging
```
Comment (skupr-anaconda) [inline recipe/meta.yaml:250]: Is `numpy` required at runtime too? See https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L88-L89

File: recipe/meta.yaml [hunk 12]
```diff
@@ -249,7 +264,7 @@ outputs:
         - wrapt >=1.11.0
         # TF-API needs to move in sync
         - tensorboard >=2.17,<2.18
-        - keras >=3.0
+        - keras >=3.2
```
Comment (skupr-anaconda) [inline recipe/meta.yaml]:
```suggestion
        - keras >=3.2.0
```
https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L118

File: recipe/meta.yaml (hunk 11 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml:250]: Do we need `pyyaml` on `linux`? Or is it the Ubuntu-specific dependency? See https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/pip_package/setup.py#L118

Comment (skupr-anaconda) [reply recipe/meta.yaml:250]: It seems python 3.11 for bazel 6 requires `lit`, see https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/ci_build/release/requirements_common.txt#L45

Comment (skupr-anaconda) [reply recipe/meta.yaml:250]: Is `certifi` required on `osx`? see https://github.com/tensorflow/tensorflow/blob/v2.17.0/tensorflow/tools/ci_build/release/requirements_common.txt#L45

Comment (cbouss) [reply recipe/meta.yaml:250]: Yes we do need numpy. It was there through transitive dependencies but I will add it explicitly. Thanks!

pyyaml is for tests only. 

We can do without certifi and lit. What the build does is pull from, for example, https://github.com/tensorflow/tensorflow/blob/v2.17.0/requirements_lock_3_11.txt

**Review by skupr-anaconda — APPROVED** [review]

**Comment by danpetry** [comment]
🎉
